src/odemis/driver/zeiss.py

```diff
@@ -0,0 +1,963 @@
+        baudrate (0<int)
+        midn (str or None): regex to match the *IDN answer
+        return:
+           (str): the name of the port used
+           (str): the identification string
+           Note: will also update ._file and ._serial
+        raises:
+            IOError: if no device are found
+        """
+        # For debugging purpose
+        if ports == "/dev/fake":
+            self._serial = RemconSimulator(timeout=1)
+            self._file = None
+            idn = self.GetVersion()
+            return ports, idn
+
+        if os.name == "nt":
+            raise NotImplementedError("Windows not supported")
+        else:
+            names = glob.glob(ports)
+        for n in names:
+            try:
+                # Ensure no one will talk to it simultaneously, and we don't talk to devices already in use
+                self._file = open(n)  # Open in RO, just to check for lock
+                try:
+                    fcntl.flock(self._file.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)  # Raises IOError if cannot lock
+                except IOError:
+                    logging.info("Port %s is busy, will not use", n)
+                    continue
+
+                self._serial = self._openSerialPort(n, baudrate)
+                try:
+                    idn = self.GetVersion()
+                except RemconError:
+                    # Can happen if the device has received some weird characters
+                    # => try again (now that it's flushed)
+                    logging.info("Device answered by an error, will try again")
+                    idn = self.GetVersion()
+                if midn and not re.search(midn, idn):
+                    logging.info("Skipping device on port %s, which identified as %s",
+                                 n, idn)
+                    continue
+                return n, idn
+            except (IOError, RemconError):
+                logging.info("Skipping device on port %s, which didn't seem to be compatible", n)
+                # not possible to use this port? next one!
+                continue
+        else:
+            raise HwError("Failed to find a device on ports '%s'. "
+                          "Check that the device is turned on and connected to "
+                          "the computer." % (ports,))
+
+    def _SendQuery(self, _cmd, timeout=1):
+        """
+        Send query to device
+        _cmd: valid query command for Remcon SEM, ending with '?'
+        """
+        cmd = _cmd + "\r"
+        with self._ser_access:
+            logging.debug("Sending command %s", cmd.encode('string_escape'))
+            self._serial.write(cmd)
+            self._serial.timeout = timeout
+
+            # Acknowledge
+            ack = self._serial.read()  # valid/invalid
+
+            # If first sign is success or fail (e.g. from previous stage move), read
+            # next character
+            if ack == RS_SUCCESS or ack == RS_FAIL:
```

Comment:
The send_query and send_order functions are not very nice because they assume that everything goes to plan and the characters can be read one-by-one. If something goes wrong or the acknowledge response is sent with a delay (as in long moves), the code runs into problems. I can improve this next week.